### ISSUE TITLE
Tracking: typed expression IR to replace string surgery

### ISSUE THREAD

**Issue by jc-macdonald** [ACCIDDA/op_system#112]
## Motivation

`_normalize.py` is **3268 lines** of regex / string-template / partial-AST surgery, and `_vectorize.py` is another 1650. Every new feature (templated aliases, repeated-axis selectors, helpers, CSE) lands as another string-position-aware pass that has to re-discover what an axis binding, a state subscript, or a sum is. The recurring symptoms:

- **#111** templated-alias bodies are inlined per LHS row by string substitution → no CSE → ~80× slowdown on contact-kernel models.
- **#88** repeated-axis selectors (`C[age, age]`, mobility `M[loc, loc]`, mutation `K[strain, strain]`) cannot be expressed because placeholder identity is coupled to axis identity at the string level.
- **#30** helpers split between compile-time string macros (`sum_over`, `integrate_over`) and runtime closures (`sum_state`, `sum_prefix`) — opaque, unserializable, untestable in isolation.
- **#22** `ExpressionString` Pydantic type was specified to expose `.ast` and `.names` precisely so we stop re-parsing strings.
- **#5** `NormalizedRhs` naming/structure can't really be cleaned up until we know what it *contains*.
- LHS-template axis values do not thread cleanly into state subscripts during alias collection (the per-loc `foi[age, loc]` case currently blocking COVID19_USA hier configs).

All of these have the same root cause: **there is no typed in-memory representation of an expression**. Every pass parses, substitutes, and re-emits strings.

## Proposal

Introduce a small typed IR (≤ ~10 node types) and rewrite passes one at a time behind a flag, with parity tests at each step.

### IR sketch

```python
# src/op_system/_ir.py
class Literal(NamedTuple):  value: float | int
class Sym(NamedTuple):      name: str                       # bare identifier
class AxisIndex(NamedTuple):
    axis: str
    coord: str | None       # None == free wildcard, else bound coord OR placeholder name
    placeholder: str | None # for #88 `$`-style independent placeholders
class Subscript(NamedTuple):
    name: str
    indices: tuple[AxisIndex, ...]
class Apply(NamedTuple):
    op: str                 # '+', '*', '/', 'pow', or registered helper name
    args: tuple["Expr", ...]
class Reduce(NamedTuple):
    kind: str               # 'apply_along' | 'sum_over' | 'integrate_over' | 'sum_state'
    bindings: tuple[tuple[str, str], ...]   # (axis, coord_var)
    body: "Expr"

Expr = Literal | Sym | Subscript | Apply | Reduce
```

### Passes (each is a tree visitor, ~30–80 lines)

1. **Parse** — `ExpressionString.ast` (#22). One Python `ast` walk → IR. No regex.
2. **Resolve helpers** (#30) — `sum_over` / `integrate_over` / `sum_state` / `sum_prefix` become `Reduce` nodes with explicit semantics. Helpers register a Python callable for codegen.
3. **Resolve axes** — every `AxisIndex` gets `axis` validated; free vs. bound vs. placeholder classified. `$`-placeholders (#88) collapse to `(axis=..., coord=None, placeholder='ap')`.
4. **Expand templates** — LHS template substitution = `expr.subs({AxisIndex(axis, None, ph): AxisIndex(axis, coord, None)})`. Per-row alias inlining = same. Threads LHS coords into state subscripts uniformly.
5. **CSE** (#111) — standard hash-cons over the IR; share subexpressions across rate cells. Trivially correct because the IR is value-typed.
6. **Vectorize / lower to JAX** — replace ~1650 lines of `_vectorize.py` with a single emit pass: `Reduce` → `jnp.sum`/`apply_along`; `Subscript` → fancy index; `Apply` → primitive op.

### Why now

We are simultaneously hitting three of the four open architecture issues on the same model (#88, #111, plus the per-loc state-subscript binding gap). Patching each with another string pass is cheaper this week and substantially more expensive every week after.

### Why not a big symbolic backend (pytensor / sympy / casadi)

Considered. They each pay for themselves only when the IR is large; ours is tiny and domain-specific (axes are first-class, not generic dims). Roll-your-own is < 500 lines, has no deps, and we can swap to `pytensor` later by re-implementing the `lower()` pass alone.

## Subtasks

- [ ] #22 land `ExpressionString` with `.ast` and `.names` — the IR's parse entry point
- [ ] [new] add `_ir.py` with the dataclasses above + a `to_ir(ast)` builder
- [ ] [new] port helper expansion (#30) to operate on IR
- [ ] [new] port template / alias substitution to IR (subsumes the LHS-template state-subscript fix below)
- [ ] [new] port repeated-axis resolution (#88) to IR — `$`-placeholders fall out
- [ ] [new] add CSE pass on IR (#111)
- [ ] [new] port `_vectorize` lowering to consume IR
- [ ] [new] make `NormalizedRhs` (#5) hold IR not strings; rename
- [ ] [new] cache + pickle `CompiledRhs` (now possible — #30 closure problem goes away)

## Non-goals

- New user-facing surface beyond #88's `$`-placeholders.
- Replacing the YAML reader.
- Performance work beyond CSE — no fusion / no kernel scheduling in this milestone.

## Acceptance

- All 214 existing tests pass against the IR-backed path.
- Bench (`postprocessing/run_hier_bench.sh` in COVID19_USA) recovers to baseline (discrete ≈ 135 ms/draw CPU, continuum ≈ 750 ms/draw GPU).
- `CompiledRhs` is picklable.

## Related

- Milestone #1 Pydantic-ify (the IR is what `ExpressionString` parses *into*)
- Milestone #4 Core Internals
- Issues #5, #22, #30, #88, #111
